### PR TITLE
Add support for parsing Gsym inlined function information

### PR DESCRIPTION
**Pull request by danielocfb**
Please see individual commits for descriptions.

### CODE REVIEW

**Comment by danielocfb** [comment]
The PR is in a reasonable state with respect to what it covers. However, what is still lacking is the higher level plumbing for `blazesym::symbolize::Symbolizer` to actually report inlined function information. I think we may want to take this as an opportunity to rework the return type of the `symbolize` function to begin with, which is why I left it it as a draft for now while I work on that.
That being said, happy to receive feedback and most of the stuff that *is* out is unlikely to change all that much, I'd guess, it's only that more will be added.

**Comment by danielocfb** [comment]
Also linking https://github.com/libbpf/blazesym/issues/192 here, which this PR will address partly.

File: tests/blazesym.rs [hunk 1]
```diff
@@ -134,6 +134,49 @@ fn symbolize_elf_dwarf_gsym() {
...
+            .enable_inlined_fns(inlined_fns)
+            .build();
+        let results = symbolizer
+            .symbolize(&src, &[0x200020a])
+            .unwrap()
+            .into_iter()
+            .collect::<Vec<_>>();
+        assert_eq!(results.len(), 1);
+
+        let (result, _addr_idx) = &results[0];
```
Comment (anakryiko) [inline tests/blazesym.rs:153]: can you please also check the non-inlined symbol name and code_info here, regardless of `inlined_fns`?

File: src/gsym/inline.rs [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+
+impl InlineInfo {
+    /// Parse Gsym `InlineInfo` from raw bytes.
+    pub fn parse(
+        data: &mut &[u8],
+        base_addr: u64,
+        lookup_addr: Option<u64>,
+    ) -> Result<Option<InlineInfo>> {
+        let range_cnt = data
+            .read_u128_leb128()
```
[the remark this answers is not in the document]
Comment (d-e-s-o) [reply src/gsym/inline.rs:26]: Yes we should do that. Given that it's basically pre-existing issue, will do it as a follow up.

File: tests/blazesym.rs (hunk 1 above)
Comment (d-e-s-o) [reply tests/blazesym.rs:153]: Will do.

File: src/gsym/resolver.rs [hunk 3]
```diff
@@ -339,13 +339,41 @@ mod tests {
...
+        // always fall into the inlined region, no matter toolchain. If not, add
+        // padding bytes/dummy instructions and adjust some more.
+        let addr = 0x200020a;
+        let syms = resolver.find_syms(addr).unwrap();
+        assert_eq!(syms.len(), 1);
+        let sym = &syms[0];
+        assert_eq!(sym.name, "factorial_inline_test");
+
+        let info = resolver.find_line_info(addr).unwrap().unwrap();
+        assert_eq!(info.direct.line, Some(21));
```
[the remark this answers is not in the document]
Comment (d-e-s-o) [reply src/gsym/resolver.rs]: Alright, fixed it up.

**Comment by d-e-s-o** [comment]
Will do the enum addition we discussed in person as a follow on. This PR is way to unwieldy already and in any case I am not going to merge it in.

File: src/gsym/resolver.rs [hunk 4]
```diff
@@ -151,76 +207,96 @@ impl SymResolver for GsymResolver<'_> {
...
+                            let second = unsafe { &mut *inlined.add(prev_i) };
+                            let () = swap(&mut first.1, &mut second.1);
+                        } else {
+                            direct_name = Some(name);
+                            if let Some(code_info) = &mut inlined[i].1 {
+                                let () = swap(code_info, &mut line_tab_info);
+                            }
+                        }
+
+                        prev = Some(i);
```
Comment (anakryiko) [inline src/gsym/resolver.rs:287]: this two-step construction seems unnecessary convoluted, tbh. Wouldn't it be simpler to construct final `inlined` vector on the first pass? It seems like you could avoid unsafe bits as well? E.g.,  first, from inline_stack[0] you extract name and keep it as "pending name", while code_info goes into sym.code_info (store code_info from line_table for later separately for now). Then on each item in inline_step[1..len), take code_info and combine it with last pending name, reassign pending name to be name you just extracted.

After the whole loop, take last pending name and combine it with code_info from line_table as last inlined element.

File: src/symbolize/symbolizer.rs [hunk 5]
```diff
@@ -166,6 +193,19 @@ impl Builder {
         self
     }
 
+    /// Enable/disable source code location information (line numbers,
+    /// file names etc.).
+    pub fn enable_src_location(mut self, enable: bool) -> Builder {
```
Comment (anakryiko) [inline src/symbolize/symbolizer.rs:198]: a bit of inconsistency. it's code_info and CodeInfo in Sym, but here it's called "src_location". Maybe let's pick one way of naming this?

File: src/symbolize/symbolizer.rs [hunk 6]
```diff
@@ -166,6 +193,19 @@ impl Builder {
...
+        self.src_location = enable;
+        self
+    }
+
+    /// Enable/disable inlined function reporting.
+    pub fn enable_inlined_fns(mut self, enable: bool) -> Builder {
```
Comment (anakryiko) [inline src/symbolize/symbolizer.rs:204]: it's a bit of eye sore to read "inline_fns", would "enable_inlines" be confusing? Then you can also call the field in Sym "inlines" and be consistent in naming throughout?

**Review by anakryiko — APPROVED** [review]
Looks good. This is a big PR, let's land it. We had a discussion separately about Sym and addr_idx, approach. If you decide to change it, let's do it in a separate PR. Also, if you aggree to refine the inline post-processing to be a single step, let's do it in a separate PR as well.

File: src/symbolize/symbolizer.rs (hunk 5 above)
Comment (danielocfb) [reply src/symbolize/symbolizer.rs:198]: Yeah, I noticed that. Will change it.

File: src/symbolize/symbolizer.rs (hunk 6 above)
Comment (danielocfb) [reply src/symbolize/symbolizer.rs:204]: > Then you can also call the field in Sym "inlines"

It's actually called `inlined`. And it currently is of type `InlinedFn`. I am not sure I want to shorten it to just a single word, to be honest. Similar for `code info` or `debug syms`. So if anything I'd rather change `Sym::inlined` to `Sym::inlined_fns`. I have no strong preference. The types involved basically add the necessary context from my perspective.

File: src/gsym/resolver.rs (hunk 4 above)
Comment (danielocfb) [reply src/gsym/resolver.rs:287]: Thanks for the comment. Okay, will take another look.

**Comment by danielocfb** [comment]
Going ahead with the merge as per the suggestion.

File: src/gsym/resolver.rs (hunk 4 above)
Comment (danielocfb) [reply src/gsym/resolver.rs:287]: https://github.com/libbpf/blazesym/pull/328

**Pull request merged**